host.go

```diff
@@ -0,0 +1,188 @@
+		return xerrors.Errorf("starting a participant: %w", err)
+	}
+
+loop:
+	for {
+		select {
+		case msg := <-h.SelfMessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, true)
+		default:
+		}
+		if err != nil {
+			break loop
+		}
+
+		select {
+		case <-h.timer.C:
+			h.log.Infof("alarm fired")
+			err = h.participant.ReceiveAlarm()
+		case msg := <-h.SelfMessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, true)
+		case msg := <-h.MessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, false)
+		case <-ctx.Done():
+			err = ctx.Err()
+		}
+		if err != nil {
+			break loop
+		}
+	}
+	h.log.Errorf("gpbfthost exiting: %+v", err)
+	return err
+}
+
+//
+// internal APIs after
+//
+
+// Returns inputs to the next GPBFT instance.
+// These are:
+// - the EC chain to propose,
+// - the power table specifying the participants,
+// - the beacon value for generating tickets.
+// These will be used as input to a subsequent instance of the protocol.
+// The chain should be a suffix of the last chain notified to the host via
+// ReceiveDecision (or known to be final via some other channel).
+func (h *gpbfthost) GetCanonicalChain() (chain gpbft.ECChain, power gpbft.PowerTable, beacon []byte) {
+	// TODO: this is just a complete fake
+	ts := sim.NewTipSetGenerator(1)
+	chain, err := gpbft.NewChain(gpbft.TipSet{Epoch: 0, Key: ts.Sample()}, gpbft.TipSet{Epoch: 1, Key: ts.Sample()})
+	if err != nil {
+		h.log.Errorf("creating chain: %+v", err)
+		return nil, *gpbft.NewPowerTable(), nil
+	}
+	table := gpbft.NewPowerTable()
+	err = table.Add(h.manifest.InitialPowerTable...)
+	if err != nil {
+		h.log.Errorf("creating powertable: %+v", err)
+		return nil, *gpbft.NewPowerTable(), nil
+	}
+
+	return chain, *table, []byte{'A'}
+}
+
+// Returns the network's name (for signature separation)
+func (h *gpbfthost) NetworkName() gpbft.NetworkName {
+	return h.manifest.NetworkName
+}
+
+// Sends a message to all other participants.
+// The message's sender must be one that the network interface can sign on behalf of.
+func (h *gpbfthost) Broadcast(msg *gpbft.GMessage) {
+	h.log.Info("broadcast")
+
+	var bw bytes.Buffer
+	err := msg.MarshalCBOR(&bw)
+	if err != nil {
+		h.log.Errorf("marshalling GMessage: %+v", err)
+	}
+	err = h.broadcast(h.runningCtx, bw.Bytes())
+	if err != nil {
+		h.log.Errorf("broadcasting GMessage: %+v", err)
+	}
+}
+
+// Returns the current network time.
+func (h *gpbfthost) Time() time.Time {
+	return time.Now()
+}
+
+// Sets an alarm to fire after the given timestamp.
+// At most one alarm can be set at a time.
+// Setting an alarm replaces any previous alarm that has not yet fired.
+// The timestamp may be in the past, in which case the alarm will fire as soon as possible
+// (but not synchronously).
+func (h *gpbfthost) SetAlarm(at time.Time) {
+	h.log.Infof("set alarm for %v", at)
+	// we cannot reuse the timer because we don't know if it was read or not
+	h.timer.Stop()
+	h.timer = time.NewTimer(time.Until(at))
+}
```

Comment:
It is the 2nd case; it is only called by the Participant. We still really can't reuse the timer because we don't know if we are in participant because the timer fired and was read, or the timer fired while the participant was running.

Reply:
I see.... I'd still consider sending a message on a control channel isntead.

Reply:
The Participant is inherently single-threaded, and we don't currently have a control channel. I can add them if you think it warrants the complexity.